### PR TITLE
4948 - ids colorpicker wc

### PR DESCRIPTION
**Pull request by danielortiz1982**
**Explain the details for making this change. What existing problem does the pull request solve?**
Porting and converting the enterprise jQuery color-picker component into an IDS Web Component. Creating a custom element web component with options to customize the color-picker, labels and the ability to define different colors with a child color web component. 

**Related github/jira issue (required)**:
Closes https://github.com/infor-design/enterprise/issues/4948

**Steps necessary to review your pull request (required)**:
- Checkout the feature branch - _**git checkout 4948-ids-colorpicker-wc**_
- Pull the latest changes - _**git pull -f**_
- Create a sample html document inside of the ids-colorpicker folder - _**./app/ids-colorpicker**_
- Create the custom element by using the markup - 
```
<ids-colorpicker>
   <color hex="#000000"></color>
   <color hex="#cccccc"></color>
   <color hex="#333333"></color>
</ids-colorpicker>
```
- Navigate to the color-picker wc url - http://localhost:4300/ids-colorpicker

**Pending completion on this Pull Request**:
- Default values
- e2e fn testing
- Percy testing
- ReadMe Docs

**Included in this Pull Request**:
- [ ] An e2e or functional test for the bug or feature.
- [ ] A note to the change log.

### CODE REVIEW

**Comment by tmcconechy** [comment]
Before we review can you
- update the PR template (make sure to include what isnt done yet and anything you need us to help look at)
- can also start by fixing the checks...

BTW converted to WIP

File: src/ids-color/ids-color.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * @type {IdsColor}
+ * @inherits IdsElement
+ * @mixes IdsEventsMixin
+ * @mixes IdsKeyboardMixin
+ */
+
+ @customElement('ids-color')
+ @scss(styles)
+
+ class IdsColor extends mix(IdsElement).with(IdsEventsMixin, IdsKeyboardMixin) {
```
Comment (tmcconechy) [inline src/ids-color/ids-color.js:25]: Probably dont need these mixins.

File: src/ids-color/ids-color.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+     constructor(){
+         super()
+     }
+
+     connectedCallback() {
```
Comment (tmcconechy) [inline src/ids-color/ids-color.js:30]: Possibly dont need

File: src/ids-colorpicker/ids-colorpicker.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import { IdsEventsMixin } from '../ids-base/ids-events-mixin';
+import { IdsKeyboardMixin } from '../ids-base/ids-keyboard-mixin';
+import '../ids-color/ids-color';
+
+// @ts-ignore
+import styles from './ids-colorpicker.scss';
+
+/**
+ * IDS Colorpick
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:16]:
```suggestion
 * IDS Color Picker
```

File: src/ids-colorpicker/ids-colorpicker.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ * @type {IdsColorpicker}
+ * @inherits IdsElement
+ * @mixes IdsEventsMixin
+ * @mixes IdsKeyboardMixin
+ */
+
+ @customElement('ids-colorpicker')
+ @scss(styles)
+
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:25]: Extra space (NIT)

File: src/ids-colorpicker/ids-colorpicker.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+ class IdsColorpicker extends mix(IdsElement).with(IdsEventsMixin, IdsKeyboardMixin) {
+     constructor(){
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:27]:
```suggestion
     constructor() {
```
(nit)

File: src/ids-colorpicker/ids-colorpicker.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+         super()
+     }
+
+     connectedCallback() {
+         this.colorPicker()
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:32]: Might need super() here. see other components

File: src/ids-colorpicker/ids-colorpicker.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+     */
+    template() {
+        return `
+        <div id="ids-colorpicker">
+            <div class="colorpicker">
+                <div class="colorpicker-container">
+                    <span class="color-preview">
+                        <input class="color-input" type="color">
+                    </span>
+                    <input type="text" class="color-input-value" value="#000000">
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:47]: Maybe set this to the value attribute

File: src/ids-colorpicker/ids-colorpicker.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                <div class="color-container hide-color-container">
+                    <slot></slot>
+                </div>
+            </div>
+        </div>
+    `;
+    }
+
+    /**
+     * Retuns IdsColorpicker shadowRoot
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:61]: Miss places comment?

File: src/ids-colorpicker/ids-colorpicker.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+     * @returns {HTMLCollection} shadowRoot
+     */
+    idsColorPicker = this.shadowRoot
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:64]: Do these need const?

File: src/ids-colorpicker/ids-colorpicker.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    colorpickerContainer = this.idsColorPicker.querySelector('#ids-colorpicker')
+    colorContainer = this.idsColorPicker.querySelector('.color-container')
+    colorpickerInput = this.idsColorPicker.querySelector('.color-input')
+    colorInputValue = this.idsColorPicker.querySelector('.color-input-value')
+    colorPreview = this.idsColorPicker.querySelector('.color-preview')
+    idsColorsArr = document.querySelectorAll('ids-color')
+
+    colorPicker() {
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:72]: Maybe a better name? Not sure what this does

File: src/ids-colorpicker/ids-colorpicker.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        // @ts-ignore
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:73]: When you merge mast on you wont need any ts-ignores anymore

File: src/ids-colorpicker/ids-colorpicker.js [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        this.idsColorsArr.forEach((element) => element.style.backgroundColor = element.getAttribute('hex'))
+        this.onEvent('click', this.colorpickerContainer, (/** @type {{ target: any; }} */ event) => {
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:75]:
```suggestion
        this.onEvent('click', this.colorpickerContainer, (event) => {
```

Might not need this anymore (multiple in document)

File: src/ids-colorpicker/ids-colorpicker.js [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+            const target = event.target
+            const openColorCondition = (target.classList.contains('colorpicker-icon') || target.classList.contains('ids-dropdown'))
+            if(openColorCondition){
```
Comment (tmcconechy) [inline src/ids-colorpicker/ids-colorpicker.js:78]:
```suggestion
            if (openColorCondition){
```

Spaces after the if's  (multiple in document)

**Review by tmcconechy — COMMENTED** [review]
I know this is WIP but adding some initial suggestions. And in addition to the inline stuff in order of importance

1. The styling is not what it was https://master-enterprise.demo.design.infor.com/components/colorpicker/example-index.html
1. Try to use ids-trigger field so we inherit from that
1. make sure to add themes see notes on https://github.com/infor-design/enterprise-wc/tree/main/src/ids-base#ids-theme-mixin (would do that when done)
1. Make sure to look at other examples we made as this seems quite inconsistent (check out tags and inputs and maybe textarea)
1. ids-color has no props -> i would think it would at least have a color prop
1. ids-colorpicker has no props -> i would think it would have all the properties of trigger fields and inputs (well it must)
1. i like the thing you did when clicking the swatch but it wasnt required. so would suggest you put it as a setting that changes from the way it currently works to use that instead and not have two ways to do it at once. 
1. Need keyboard/labels/focus to work
1. some tests and lint is failing (but you mentioned it wasnt working yet)

**Comment by danielortiz1982** [comment]
Thanks for the feedback @tmcconechy

**Comment by tmcconechy** [comment]
Closing this for now - until its ready.. Then we can open it back up again 👍🏻 
Comments still here for reference

**Comment by danielortiz1982** [comment]
> Closing this for now - until its ready.. Then we can open it back up again 👍🏻
> Comments still here for reference

Cool beans! Beans are cool. Thnx @tmcconechy :)

**Pull request closed without merging**